### ISSUE TITLE
How to support dynamic graph on the server

### ISSUE THREAD

**Issue by bendecoste** [OpenMined/OpenMined#221]
The goal for the openmined client (pytorch) api is to mirror frameworks that users will already be familiar with.  e.g. we should support

```
openmined
  keras
    <mimic keras api>
  pytorch
    <mimic pytorch api>
```

and we are currently focusing on PyTorch.  A big feature in pytorch is the dynamic compute graph which is calculated each run in the `forward` function (more info [here](http://pytorch.org/about/#dynamic-neural-networks-tape-based-autograd)).  This discussion is about how do we support this feature in an optimal way.

### State of the union

This feature is actually supported right now.  But it is far from optimal.  To support it, you must implement your models client side and each pass of the forward function will make as many calls as there are layers (or more).  This means to train a serious model you can be *easily* getting into the millions of socket messages.

This isn't a problem for static models where there is a clear path of data at setup time (e.g. keras style / pytorch sequential)

I have been talking with @iamtrask about this quite a bit but I wanted to open up the conversation to more people!

**Comment by emregoynugur** [comment]
We, previously, discussed that at some point we have to convert/serialize the Python code into a format that the Unity application would understand. i.e. data providers will just accept model requests from data scientists, but they will not run any Python code. For this to happen, data providers' Unity application should be able to download the corresponding Unity model. 

Maybe it is time for us to seriously consider this. We could allow data scientists to create execution graphs (composition of ops) and get its final result from the backend. That way data scientists could use our Python API like PyTorch and reduce the number of socket calls, when he/she sees fit. We can also serialize the entire model in the same fashion.

**Comment by iamtrask** [comment]
Been thinking about this a lot lately - I think that we can take very flexible inspiration from this class.

https://github.com/OpenMined/OpenMined/blob/master/UnityProject/Assets/OpenMined/Syft/Model/Sequential.cs

This is a static graph - but it uses a dynamic graph under the hood. I think that we could probably pre-define more and more general/abstract graphs that are specifiable on the Python side and serializable to the C# side. In the short run, expanding these is perhaps the highest priority.

Long run, the most difficult piece for dynamic computation graphs is when you generate indices (perhaps by taking .max() from a softmax) and then use those to index into a subset of rows in a weight matrix. However, even though the graph is dynamic (uses unpredicable subsets of weights at each forward propagation), the sortof "instruction graph" needed for serialization is still very well defined. I think this is very instructive as it means that, for the purpose of serialization/deserialization and Python/C# interop.

Slightly different but related is the use of if statements directly in the Python code... this is quite difficult but I believe @bendecoste has an idea for this... where basically we treat an if statement like a graph call.... in the neural network sense... it treats an "if" statement like a kind of layer type itself which I think is pretty clever. In that sense, it collapses this issue into simply solving the one in previous paragraph.

All that is to say... perhaps the roadmap is something like:
- build the standard graph types... Sequential, Recurrent (LSTM), and Sibling network graphs. Build serialization and such... aka... start simple.
- implement index level autograd (like when you select random subsets of rows)... we don't currently have this.
- experiment with if/else layer types
.... regroup and discuss... we'll probably have more information at this point on what will likely be successful.

The good news is that the Pareto principle is strong here - most people will use Sequential and Recurrent for almost everything.
